Source/WebCore/page/Navigation.cpp

```diff
@@ -717,48 +762,72 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-            else
-                failure = true;
-            // FIXME: We need to keep around the failure reason but the generated handleEvent() catches and consumes it.
+            else if (callbackResult.type() == CallbackResultType::ExceptionThrown) {
+                // FIXME: We need to keep around the failure reason but the generated handleEvent() catches and consumes it.
+                auto promiseAndWrapper = createPromiseAndWrapper(*document);
+                promiseAndWrapper.second->reject(ExceptionCode::TypeError);
+                promiseList.append(WTFMove(promiseAndWrapper.first));
+            }
         }
 
-        // FIXME: Step 33.4: We need to wait for all promises.
+        if (promiseList.isEmpty()) {
+            auto promiseAndWrapper = createPromiseAndWrapper(*document);
+            promiseAndWrapper.second->resolveWithCallback([](JSDOMGlobalObject&) {
+                return JSC::jsUndefined();
+            });
+            promiseList.append(WTFMove(promiseAndWrapper.first));
+        }
 
-        if (document->isFullyActive() && !abortController->signal().aborted()) {
-            // If a new event has been dispatched in our event handler then we were aborted above.
-            if (m_ongoingNavigateEvent != event.ptr())
-                return false;
+        waitForAllPromises(promiseList, [=, this]() {
+            if (abortController->signal().aborted() || !document->isFullyActive() || !m_ongoingNavigateEvent)
+                return;
 
+            m_ongoingNavigateEvent->finish();
             m_ongoingNavigateEvent = nullptr;
 
-            event->finish();
+            dispatchEvent(Event::create(eventNames().navigatesuccessEvent, { }));
 
-            if (!failure) {
-                dispatchEvent(Event::create(eventNames().navigatesuccessEvent, { }));
+            if (RefPtr transition = std::exchange(m_transition, nullptr))
+                transition->resolvePromise();
 
-                if (RefPtr transition = std::exchange(m_transition, nullptr))
-                    transition->resolvePromise();
+            m_ongoingNavigateEvent = nullptr;
 
-                if (apiMethodTracker)
-                    resolveFinishedPromise(apiMethodTracker.get());
-            } else {
-                // FIXME: Fill in error information with exception from promise calls above.
-                auto exception = Exception(ExceptionCode::UnknownError);
-                dispatchEvent(ErrorEvent::create(eventNames().navigateerrorEvent, exception.message(), { }, 0, 0, { }));
+            if (apiMethodTracker)
+                resolveFinishedPromise(apiMethodTracker.get());
 
-                if (RefPtr transition = std::exchange(m_transition, nullptr))
-                    transition->rejectPromise(exception);
-            }
-        } else {
-            // FIXME: and the following failure steps given reason rejectionReason:
+        }, [=, this](JSC::JSValue result) {
```

Comment:
How do we know that this has not been destroyed when the lambda is called?

Reply:
Added WeakRef.